File: 48. Rotate Image/solution.go
func rotate(matrix [][]int) {
    N := len(matrix) - 1
    for i:=0; i<len(matrix)/2; i++ {
        upperLimit := N-i
        lowerLimit := i

        for j:=lowerLimit; j < upperLimit; j++ {
            diff := j - i
            tmp := matrix[i][j]
            matrix[i][j] = matrix[upperLimit-diff][i]
            matrix[upperLimit-diff][i] = matrix[upperLimit][upperLimit-diff]
            matrix[upperLimit][upperLimit-diff] = matrix[j][upperLimit]
            matrix[j][upperLimit] = tmp
        }
    }
}

File: 48. Rotate Image/solution.py
class Solution:
    def rotate(self, matrix: 'List[List[int]]') -> 'None':
        """
        Do not return anything, modify matrix in-place instead.
        """
        # edge case
        n = len(matrix)
        if n < 2:
            return
        
        # transpose the matrix
        for i in range(n):
            for j in range(i+1, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
                
        # reverse the rows
        for i in range(n):
            j = 0
            k = n-1
            while j < k:
                matrix[i][j], matrix[i][k] = matrix[i][k], matrix[i][j]
                j += 1
                k -= 1
